extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProvider.java

```diff
@@ -387,22 +392,46 @@ public TokenIntrospection apply(TokenIntrospection introspectionResult, Throwabl
+                                final String requiredClaimName = requiredClaim.getKey();
+                                if (!introspectionResult.contains(requiredClaimName)) {
+                                    LOG.debugf("Introspection claim %s is missing", requiredClaimName);
                                     throw new AuthenticationFailedException(tokenMap(token, idToken));
                                 }
-                                if (introspectionClaimValue == null) {
-                                    LOG.debugf("Introspection claim %s is missing", requiredClaim.getKey());
+                                final Set<String> requiredClaimValues = requiredClaim.getValue();
+                                if (requiredClaimValues.size() == 1) {
+                                    String introspectionClaimValue = null;
+                                    try {
+                                        introspectionClaimValue = introspectionResult.getString(requiredClaimName);
+                                    } catch (ClassCastException ex) {
+                                        LOG.debugf("Introspection claim %s is not String", requiredClaimName);
+                                    }
+                                    String requiredClaimValue = requiredClaimValues.iterator().next();
+                                    if (requiredClaimValue.equals(introspectionClaimValue)) {
+                                        continue;
+                                    }
+                                }
+                                final JsonArray actualClaimValueArr;
+                                try {
+                                    actualClaimValueArr = requireNonNull(introspectionResult.getArray(requiredClaimName));
+                                } catch (Exception ignored) {
+                                    LOG.debugf("Introspection claim %s is neither string or array", requiredClaimName);
                                     throw new AuthenticationFailedException(tokenMap(token, idToken));
                                 }
-                                if (!introspectionClaimValue.equals(requiredClaim.getValue())) {
+                                requiredClaimValuesLoop: for (String requiredClaimValue : requiredClaimValues) {
+                                    for (int i = 0; i < actualClaimValueArr.size(); i++) {
+                                        try {
+                                            String actualClaimValue = actualClaimValueArr.getString(i);
+                                            if (requiredClaimValue.equals(actualClaimValue)) {
+                                                continue requiredClaimValuesLoop;
+                                            }
+                                        } catch (Exception ignored) {
+                                            // try next actual claim value
```

Comment:
Should we fail instead ? I see my own code above which just logs a message if there is a class cast exception - we should probably fail there too

Reply:
> Should we fail instead ? I see my own code above which just logs a message if there is a class cast exception - we should probably fail there too

Behavior you set didn't change, but you were comparing _x_ to _x_ (one to one), while I am comparing_many_ against _many_ and requested behavior is that _expected claim values_ are subset of _actual claim values_, so it is alright to try another _actual claim value item_. The failure happens little later.

Reply:
More specifically, it fails on the line 435.